From 8375202bfb13a7c76d31ab6c2cc8402eae87135e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:02:09 -0400
Subject: [usb] xhci: Set L1 device slot on USB2 LPM enable/disable

Message-id: <1410887463-20674-58-git-send-email-dzickus@redhat.com>
Patchwork-id: 95163
O-Subject: [RHEL7 PATCH 057/591] xhci: Set L1 device slot on USB2 LPM enable/disable.
Bugzilla: 1032371
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1032371

commit 58e21f73975ec927119370635bf68b9023831c56
Author: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Date:   Mon Oct 7 17:17:20 2013 -0700

    xhci: Set L1 device slot on USB2 LPM enable/disable.

    To enable USB 2.0 Link Power Management (LPM), the xHCI host controller
    needs the device slot ID to generate the device address used in L1 entry
    tokens.  That information is set in the L1 device slot ID field of the
    USB 2.0 LPM registers.

    Currently, the L1 device slot ID is overwritten when the xHCI driver
    initiates the software test of USB 2.0 Link PM in
    xhci_usb2_software_lpm_test.  It is never cleared when USB 2.0 Link PM
    is disabled for the device.  That should be harmless, because the
    Hardware LPM Enable (HLE) bit is cleared when USB 2.0 Link PM is
    disabled, so the host should not pay attention to the slot ID.

    This patch should have no effect on host behavior, but since
    xhci_usb2_software_lpm_test is going away in an upcoming bug fix patch,
    we need to move that code to the function that enables and disables USB
    2.0 Link PM.

    This patch should be backported to kernels as old as 3.11, that contain
    the commit a558ccdcc71c7770c5e80c926a31cfe8a3892a09 "usb: xhci: add USB2
    Link power management BESL support".  The upcoming bug fix patch is also
    marked for that stable kernel.

    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
    Cc: stable@vger.kernel.org
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index 6e0d886..b86d102 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -4228,7 +4228,7 @@ int xhci_set_usb2_hardware_lpm(struct usb_hcd *hcd,
   }
 
   pm_val &= ~PORT_HIRD_MASK;
-  pm_val |= PORT_HIRD(hird) | PORT_RWE;
+  pm_val |= PORT_HIRD(hird) | PORT_RWE | PORT_L1DS(udev->slot_id);
   xhci_writel(xhci, pm_val, pm_addr);
   pm_val = xhci_readl(xhci, pm_addr);
   pm_val |= PORT_HLE;
@@ -4236,7 +4236,7 @@ int xhci_set_usb2_hardware_lpm(struct usb_hcd *hcd,
   /* flush write */
   xhci_readl(xhci, pm_addr);
  } else {
-  pm_val &= ~(PORT_HLE | PORT_RWE | PORT_HIRD_MASK);
+  pm_val &= ~(PORT_HLE | PORT_RWE | PORT_HIRD_MASK | PORT_L1DS_MASK);
   xhci_writel(xhci, pm_val, pm_addr);
   /* flush write */
   xhci_readl(xhci, pm_addr);
diff --git a/drivers/usb/host/xhci.h b/drivers/usb/host/xhci.h
index 941d5f5..ed3a425 100644
--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -383,6 +383,7 @@ struct xhci_op_regs {
 #define PORT_RWE  (1 << 3)
 #define PORT_HIRD(p)  (((p) & 0xf) << 4)
 #define PORT_HIRD_MASK  (0xf << 4)
+#define PORT_L1DS_MASK  (0xff << 8)
 #define PORT_L1DS(p)  (((p) & 0xff) << 8)
 #define PORT_HLE  (1 << 16)
 
-- 
1.7.1